### ISSUE TITLE
apt_key does not update after installing a new key from url

### ISSUE THREAD

**Issue by rcuza** [ansible/ansible#25091]
##### ISSUE TYPE

 - Bug Report

##### COMPONENT NAME
`apt_key_module` 

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /Users/raulcuza/dev/cg/transit/transit/infra/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, Dec 18 2016, 07:03:34) [GCC 4.2.1 Compatible Apple LLVM 7.0.2 (clang-700.1.81)]
```

##### CONFIGURATION

```
[defaults]
ansible_managed = ixntransit ansible managed: {file} modified on %Y-%m-%d %H:%M:%S by {uid} on {host}
inventory = inventories/dev
roles_path = ./roles
retry_files_enabled = False
transport = ssh
vault_password_file = ~/.vault_pass

[ssh_connection]
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m
control_path = ~/.ssh/ansible-%%r@%%h:%%p
pipelining = True
```

##### OS / ENVIRONMENT

ubuntu 16.04

##### SUMMARY

When I run a playbook that installs a new key, it is not able to install packages from the repo that is also added. If I manually ssh into the box and run ~`sudo apt-key update`~ `sudo apt-get update`, I can then run the playbook successfully.

##### STEPS TO REPRODUCE

* Start with a stock Ubuntu 16.04 with only `sudo apt-get install -y python python-apt` run on it.
* Run the below role against it.

```yaml
---
- name: install esl-erlang apt repo
  apt_repository:
    filename: erlang-solutions
    repo: deb http://binaries.erlang-solutions.com/debian xenial contrib
    state: present
    update_cache: yes

- name: install esl-erlang apt signing key
  apt_key:
    url: "https://packages.erlang-solutions.com/ubuntu/erlang_solutions.asc"
    state: present

- name: install apt packages
  apt:
    autoremove: yes
    cache_valid_time: 300
    name: "{{ item }}"
    state: present
    update_cache: yes
  with_items:
    - esl-erlang
    - elixir
```

##### EXPECTED RESULTS

I expect all the packages from the new repo to be installed with authentication verified.

##### ACTUAL RESULTS

I get an error stating that the packages cannot be installed because they are not verified.

```
TASK [eucalyptus : install apt packages] *******************************************************************************************************************************************
failed: [10.0.0.12] (item=[u'esl-erlang', u'elixir', u'libssl1.0.0', u'locales', u'postgresql-client']) => {"cache_update_time": 1495847381, "cache_updated": false, "failed": true, "item": ["esl-erlang", "elixir", "libssl1.0.0", "locales", "postgresql-client"], "msg": "'/usr/bin/apt-get -y -o \"Dpkg::Options::=--force-confdef\" -o \"Dpkg::Options::=--force-confold\"   --auto-remove  install 'esl-erlang' 'elixir' 'postgresql-client'' failed: E: There were unauthenticated packages and -y was used without --allow-unauthenticated\n", "stderr": "E: There were unauthenticated packages and -y was used without --allow-unauthenticated\n", "stderr_lines": ["E: There were unauthenticated packages and -y was used without --allow-unauthenticated"], "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following additional packages will be installed:\n  adwaita-icon-theme at-spi2-core colord colord-data dconf-gsettings-backend\n  dconf-service emacsen-common erlang-mode fontconfig fontconfig-config\n  fonts-dejavu-core glib-networking glib-networking-common\n  glib-networking-services gsettings-desktop-schemas hicolor-icon-theme\n  humanity-icon-theme libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data\n  libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3\n  libboost-filesystem1.58.0 libboost-system1.58.0 libcairo-gobject2 libcairo2\n  libcolord2 libcolorhug2 libcroco3 libcups2 libdatrie1 libdconf1\n  libdrm-amdgpu1 libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libegl1-mesa\n  libepoxy0 libexif12 libfontconfig1 libgbm1 libgd3 libgdk-pixbuf2.0-0\n  libgdk-pixbuf2.0-common libgl1-mesa-dri libgl1-mesa-glx libglapi-mesa\n  libgphoto2-6 libgphoto2-l10n libgphoto2-port12 libgraphite2-3 libgtk-3-0\n  libgtk-3-bin libgtk-3-common libgtk2.0-0 libgtk2.0-bin libgtk2.0-common\n  libgudev-1.0-0 libgusb2 libharfbuzz0b libice6 libieee1284-3 libjbig0\n  libjpeg-turbo8 libjpeg8 libjson-glib-1.0-0 libjson-glib-1.0-common\n  liblcms2-2 libllvm3.8 libltdl7 libmirclient9 libmircommon5 libmirprotobuf3\n  libnotify4 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0\n  libpciaccess0 libpixman-1-0 libpq5 libprotobuf-lite9v5 libproxy1v5\n  librest-0.7-0 librsvg2-2 librsvg2-common libsane libsane-common libsctp1\n  libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libthai-data libthai0 libtiff5\n  libtxc-dxtn-s2tc0 libvpx3 libwayland-client0 libwayland-cursor0\n  libwayland-egl1-mesa libwayland-server0 libwxbase3.0-0v5 libwxgtk3.0-0v5\n  libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0\n  libxcb-render0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcomposite1\n  libxcursor1 libxdamage1 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxpm4\n  libxrandr2 libxrender1 libxshmfence1 libxtst6 libxxf86vm1\n  notification-daemon postgresql-client-9.5 postgresql-client-common\n  ubuntu-mono x11-common\nSuggested packages:\n  colord-sensor-argyll erlang erlang-manpages erlang-doc default-jre-headless\n  | java2-runtime-headless | java1-runtime-headless | java2-runtime\n  | java1-runtime cups-common libgd-tools gphoto2 gvfs liblcms2-utils\n  librsvg2-bin avahi-daemon hplip libsane-extras sane-utils lksctp-tools\n  postgresql-9.5 postgresql-doc-9.5\nThe following NEW packages will be installed:\n  adwaita-icon-theme at-spi2-core colord colord-data dconf-gsettings-backend\n  dconf-service elixir emacsen-common erlang-mode esl-erlang fontconfig\n  fontconfig-config fonts-dejavu-core glib-networking glib-networking-common\n  glib-networking-services gsettings-desktop-schemas hicolor-icon-theme\n  humanity-icon-theme libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data\n  libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3\n  libboost-filesystem1.58.0 libboost-system1.58.0 libcairo-gobject2 libcairo2\n  libcolord2 libcolorhug2 libcroco3 libcups2 libdatrie1 libdconf1\n  libdrm-amdgpu1 libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libegl1-mesa\n  libepoxy0 libexif12 libfontconfig1 libgbm1 libgd3 libgdk-pixbuf2.0-0\n  libgdk-pixbuf2.0-common libgl1-mesa-dri libgl1-mesa-glx libglapi-mesa\n  libgphoto2-6 libgphoto2-l10n libgphoto2-port12 libgraphite2-3 libgtk-3-0\n  libgtk-3-bin libgtk-3-common libgtk2.0-0 libgtk2.0-bin libgtk2.0-common\n  libgudev-1.0-0 libgusb2 libharfbuzz0b libice6 libieee1284-3 libjbig0\n  libjpeg-turbo8 libjpeg8 libjson-glib-1.0-0 libjson-glib-1.0-common\n  liblcms2-2 libllvm3.8 libltdl7 libmirclient9 libmircommon5 libmirprotobuf3\n  libnotify4 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0\n  libpciaccess0 libpixman-1-0 libpq5 libprotobuf-lite9v5 libproxy1v5\n  librest-0.7-0 librsvg2-2 librsvg2-common libsane libsane-common libsctp1\n  libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libthai-data libthai0 libtiff5\n  libtxc-dxtn-s2tc0 libvpx3 libwayland-client0 libwayland-cursor0\n  libwayland-egl1-mesa libwayland-server0 libwxbase3.0-0v5 libwxgtk3.0-0v5\n  libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0\n  libxcb-render0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcomposite1\n  libxcursor1 libxdamage1 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxpm4\n  libxrandr2 libxrender1 libxshmfence1 libxtst6 libxxf86vm1\n  notification-daemon postgresql-client postgresql-client-9.5\n  postgresql-client-common ubuntu-mono x11-common\n0 upgraded, 134 newly installed, 0 to remove and 19 not upgraded.\nNeed to get 84.2 MB of archives.\nAfter this operation, 285 MB of additional disk space will be used.\nWARNING: The following packages cannot be authenticated!\n  esl-erlang elixir erlang-mode\n", "stdout_lines": ["Reading package lists...", "Building dependency tree...", "Reading state information...", "The following additional packages will be installed:", "  adwaita-icon-theme at-spi2-core colord colord-data dconf-gsettings-backend", "  dconf-service emacsen-common erlang-mode fontconfig fontconfig-config", "  fonts-dejavu-core glib-networking glib-networking-common", "  glib-networking-services gsettings-desktop-schemas hicolor-icon-theme", "  humanity-icon-theme libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data", "  libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3", "  libboost-filesystem1.58.0 libboost-system1.58.0 libcairo-gobject2 libcairo2", "  libcolord2 libcolorhug2 libcroco3 libcups2 libdatrie1 libdconf1", "  libdrm-amdgpu1 libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libegl1-mesa", "  libepoxy0 libexif12 libfontconfig1 libgbm1 libgd3 libgdk-pixbuf2.0-0", "  libgdk-pixbuf2.0-common libgl1-mesa-dri libgl1-mesa-glx libglapi-mesa", "  libgphoto2-6 libgphoto2-l10n libgphoto2-port12 libgraphite2-3 libgtk-3-0", "  libgtk-3-bin libgtk-3-common libgtk2.0-0 libgtk2.0-bin libgtk2.0-common", "  libgudev-1.0-0 libgusb2 libharfbuzz0b libice6 libieee1284-3 libjbig0", "  libjpeg-turbo8 libjpeg8 libjson-glib-1.0-0 libjson-glib-1.0-common", "  liblcms2-2 libllvm3.8 libltdl7 libmirclient9 libmircommon5 libmirprotobuf3", "  libnotify4 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0", "  libpciaccess0 libpixman-1-0 libpq5 libprotobuf-lite9v5 libproxy1v5", "  librest-0.7-0 librsvg2-2 librsvg2-common libsane libsane-common libsctp1", "  libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libthai-data libthai0 libtiff5", "  libtxc-dxtn-s2tc0 libvpx3 libwayland-client0 libwayland-cursor0", "  libwayland-egl1-mesa libwayland-server0 libwxbase3.0-0v5 libwxgtk3.0-0v5", "  libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0", "  libxcb-render0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcomposite1", "  libxcursor1 libxdamage1 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxpm4", "  libxrandr2 libxrender1 libxshmfence1 libxtst6 libxxf86vm1", "  notification-daemon postgresql-client-9.5 postgresql-client-common", "  ubuntu-mono x11-common", "Suggested packages:", "  colord-sensor-argyll erlang erlang-manpages erlang-doc default-jre-headless", "  | java2-runtime-headless | java1-runtime-headless | java2-runtime", "  | java1-runtime cups-common libgd-tools gphoto2 gvfs liblcms2-utils", "  librsvg2-bin avahi-daemon hplip libsane-extras sane-utils lksctp-tools", "  postgresql-9.5 postgresql-doc-9.5", "The following NEW packages will be installed:", "  adwaita-icon-theme at-spi2-core colord colord-data dconf-gsettings-backend", "  dconf-service elixir emacsen-common erlang-mode esl-erlang fontconfig", "  fontconfig-config fonts-dejavu-core glib-networking glib-networking-common", "  glib-networking-services gsettings-desktop-schemas hicolor-icon-theme", "  humanity-icon-theme libatk-bridge2.0-0 libatk1.0-0 libatk1.0-data", "  libatspi2.0-0 libavahi-client3 libavahi-common-data libavahi-common3", "  libboost-filesystem1.58.0 libboost-system1.58.0 libcairo-gobject2 libcairo2", "  libcolord2 libcolorhug2 libcroco3 libcups2 libdatrie1 libdconf1", "  libdrm-amdgpu1 libdrm-intel1 libdrm-nouveau2 libdrm-radeon1 libegl1-mesa", "  libepoxy0 libexif12 libfontconfig1 libgbm1 libgd3 libgdk-pixbuf2.0-0", "  libgdk-pixbuf2.0-common libgl1-mesa-dri libgl1-mesa-glx libglapi-mesa", "  libgphoto2-6 libgphoto2-l10n libgphoto2-port12 libgraphite2-3 libgtk-3-0", "  libgtk-3-bin libgtk-3-common libgtk2.0-0 libgtk2.0-bin libgtk2.0-common", "  libgudev-1.0-0 libgusb2 libharfbuzz0b libice6 libieee1284-3 libjbig0", "  libjpeg-turbo8 libjpeg8 libjson-glib-1.0-0 libjson-glib-1.0-common", "  liblcms2-2 libllvm3.8 libltdl7 libmirclient9 libmircommon5 libmirprotobuf3", "  libnotify4 libpango-1.0-0 libpangocairo-1.0-0 libpangoft2-1.0-0", "  libpciaccess0 libpixman-1-0 libpq5 libprotobuf-lite9v5 libproxy1v5", "  librest-0.7-0 librsvg2-2 librsvg2-common libsane libsane-common libsctp1", "  libsm6 libsoup-gnome2.4-1 libsoup2.4-1 libthai-data libthai0 libtiff5", "  libtxc-dxtn-s2tc0 libvpx3 libwayland-client0 libwayland-cursor0", "  libwayland-egl1-mesa libwayland-server0 libwxbase3.0-0v5 libwxgtk3.0-0v5", "  libx11-xcb1 libxcb-dri2-0 libxcb-dri3-0 libxcb-glx0 libxcb-present0", "  libxcb-render0 libxcb-shm0 libxcb-sync1 libxcb-xfixes0 libxcomposite1", "  libxcursor1 libxdamage1 libxfixes3 libxi6 libxinerama1 libxkbcommon0 libxpm4", "  libxrandr2 libxrender1 libxshmfence1 libxtst6 libxxf86vm1", "  notification-daemon postgresql-client postgresql-client-9.5", "  postgresql-client-common ubuntu-mono x11-common", "0 upgraded, 134 newly installed, 0 to remove and 19 not upgraded.", "Need to get 84.2 MB of archives.", "After this operation, 285 MB of additional disk space will be used.", "WARNING: The following packages cannot be authenticated!", "  esl-erlang elixir erlang-mode"]}
```

**Comment by rcuza** [comment]
This bug appears to be related to #25070.

Adding the following after installing the key lets me work around the problem:
```
- name: manually update apt-key until bus is fixed
  shell: apt-get update
  become: true
  become_user: root
```
**Update**: I was originally thinking the problem was related to https://askubuntu.com/questions/75565/why-am-i-getting-authentication-errors-for-packages-from-an-ubuntu-repository#75575 but it turns out that link solves a different problem. 

The problem I found is that an Ansible user must be knowledgeable enough to install signing keys and then run an apt cache update. If this is the case, the `apt_key` module should have an option to update the apt cache when it is run. 

Another way to solve this is to give the `apt_repository` module the ability to add a signing key for a new repo. I'm not sure if this fits the model Ansible is using behind the scene but it would make for a nicer user interface.

**Comment by ansibot** [comment]
cc @jvantuyl
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by rcuza** [comment]
I did some further tests and it looks like the problem is my order of operations. Maybe instead of a bug this is really a feature request. 

By adding the repo with an update cache operation, then adding the signing key, then installing the packages with a `cache_valid_time` option, I prevented the `apt` module from running `apt-get update` before trying to install the packages signed with the new signing key. 

Changing the order of operations (i.e. install key first, then repo) also solves the problem because it means a cache update is run after the key is installed.

Additional details about the tests: 
* Test were run using Ubuntu's AWS image ami-45224425 in us-west-2.
* Using `raw` means the old workaround runs every time which is not ideal.
